contributors/design-proposals/network/pod-resolv-conf.md

```diff
@@ -0,0 +1,167 @@
+
+It is also possible for the user to insert a short shell script snippet that rewrites `resolv.conf` on container start-up. This has the same problems as the previous approach and is also awkward for the user. This also forces the container to have additional executable code such as a shell or scripting engine which increases the applications security surface area.
+
+
+# Proposal sketch
+
+Add a new `pod.dnsPolicy: Custom` that allows for user customization of `resolv.conf`.
+
+
+## Pod API example
+
+In the example below, the user wishes to add the pod namespace and a custom expansion to the search path, as they do not use the other name aliases:
+
+```yaml
+# Pod spec
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: ns1
+  name: example
+spec:
+  containers:
+  - name: example
+    image: example
+  dnsPolicy: Custom
+  dnsParams:
+    searchPaths:
+    - $NAMESPACE.svc.$CLUSTER
+    - my.dns.search.suffix
+    - $HOST
+    options:
+    - ndots: 2
+```
+
+Given the following host `/etc/resolv.conf`:
+
+```bash
+nameserver 1.2.3.4
+search foo.com
+options: ndots: 1
+```
+
+The pod will get the following `/etc/resolv.conf`:
+
+```bash
+nameserver 10.240.0.10
+# Populated from searchPaths
+search ns1.svc.cluster.local my.dns.search.suffix foo.com
+# Populated from options
+options ndots: 2
+```
+
+## More examples
+
+The following is a Pod dnsParams that only contains the host search paths:
+
+```yaml
+dnsParams:
+  searchPaths:
+  - $HOST
+```
+
+Override `ndots` and add custom search path. Note that overriding the ndot may break the functionality of some of the search paths the
+
+```yaml
+dnsParams:
+  searchPaths:
+  - my.custom.suffix
+  - $HOST
+  options:
+  - ndots: 3
+```
+
+# API changes
+
+```go
+type PodSpec struct {
+	...
+	DNSPolicy string
+	DNSParams *PodDNSParams
+}
+
+type PodDNSParams struct {
+	SearchPaths: []string
+	Options: []string
+}
+
+// This will not appear in types.go but is here for explication purposes.
+type DNSParamsSubstitution string
+const (
+	DNSParamsSearchPathNamespace = "$NAMESPACE"
+	DNSParamsSearchPathClusterDomain = "$CLUSTER"
+	DNSParamsSearchPathHostPaths = "$HOST"
+)
+```
+
+## Semantics
+### searchPath
+If `dnsPolicy: Custom` is used, then the `search` line will be constructed from the entries listed in `dnsParams.searchPath`:
+```go
+// Note: pseudocode does not include input validation.
+func SearchPath(params *PodDNSParams) []string {
+	var searchPaths []string
+	for _, entry := range params.SearchPaths {
+		if entry == "$HOST" {
+			searchPaths = append(searchPaths, HostSearchPaths...)
+			break
+		}
+		searchPaths = append(searchPaths, expand(entry))
+	}
+	return searchPaths
+}
+
+func expand(s string) string {
+	labels := strings.Split(s, ".")
+	for i := range labels {
+		if labels[0] == "$" {
+			labels[i] = Substitute(labels[i])
+		}
+	}
+	return strings.Join(labels, ".")
+}
+```
+
+#### Substitutions
+`Substitute` will replace labels that begin with `$` with values for the given Pod:
```

Comment:
is $(NAMESPACE)12345 going to => foobar12345 or is it an error?

Reply:
updated, the subst will be label for label.